File: code_snippets/kotlin/src/main/kotlin/usecases/microservices/service/SubscriptionService.kt
package usecases.microservices.service

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import usecases.microservices.SubscriptionRequest

// <start_here>
// <mark_1>
@Service
class SubscriptionService {
  // </mark_1>

  // <mark_1>
  @Handler
  suspend fun add(ctx: Context, req: SubscriptionRequest) {
    // </mark_1>
    val paymentId = ctx.random().nextUUID().toString()

    // <mark_2>
    val payRef = ctx.runBlock { createRecurringPayment(req.creditCard, paymentId) }
    // </mark_2>

    for (subscription in req.subscriptions) {
      // <mark_2>
      ctx.runBlock { createSubscription(req.userId, subscription, payRef) }
      // </mark_2>
    }
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(SubscriptionService()) })
}
// <end_here>

fun createSubscription(userId: String, subscription: String, paymentRef: String): String {
  return "SUCCESS"
}

fun createRecurringPayment(creditCard: String, paymentId: String): String {
  return "payment-reference"
}

File: code_snippets/go/usecases/microservices/service/subscriptionservice.go
package service

import (
	"context"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
)

type SubscriptionRequest struct {
	UserID        string   `json:"userId"`
	CreditCard    string   `json:"creditCard"`
	Subscriptions []string `json:"subscriptions"`
}

// <start_here>
// <mark_1>
type SubscriptionService struct{}

func (SubscriptionService) Add(ctx restate.Context, req SubscriptionRequest) error {
	// </mark_1>
	paymentId := restate.Rand(ctx).UUID().String()
	// <mark_2>

	payRef, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return CreateRecurringPayment(req.CreditCard, paymentId)
	})
	// </mark_2>
	if err != nil {
		return err
	}

	for _, subscription := range req.Subscriptions {
		// <mark_2>
		if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
			return restate.Void{}, CreateSubscription(req.UserID, subscription, payRef)
			// </mark_2>
		}); err != nil {
			return err
		}
	}

	return nil
}

func main() {
	if err := server.NewRestate().
		Bind(restate.Reflect(SubscriptionService{})).
		Start(context.Background(), ":9080"); err != nil {
		log.Fatal(err)
	}
}

// <end_here>

func CreateSubscription(id string, subscription string, ref string) error {
	return nil
}

func CreateRecurringPayment(card string, id string) (string, error) {
	return id, nil
}
